### PR TITLE
feat(homepage): add in new draggable components

### PR DESCRIPTION
**Pull request by seaerchin**
**This is marked as draft due to some visual inconsistencies/clarification required/undone sections but the remaining stuff should be minor**
## Problem
`EditHomepage` has messy code, which makes it difficult to attempt to refactor the styling. The code itself is messy due to render logic being placed together wtih dnd logic, as well as components taking a huge number of props due to ill-defined responsibilities. This refactor does quite a few things in an attempt to tackle said messy code, which will be explained below.

## Solution
1. This shifts dnd responsibility over to `Editable.Draggable` - this component provides the `id/type` (which are required by the library) through the use of the `draggableId` prop. This is an **intermediate** measure to ensure compat whilst refactoring - we could have a better abstraction here but due to time constraints, the pragmatic choice was to just pass in `draggableId` initially.
2. Styling for various cards are provided by `<Editable | Draggable>AccordionItem` - the choice was made to split into 2 components because of the semantic difference (1 is draggable, 1 isn't) + a pragmatic reason that `Draggable` requires a render prop `(draggableProvided) => <MyComponent />`. Additionally, this component **does not** provide an abstraction over `AccordionItem`, which it should (ie, be split into either smaller pieces and/or accept `AccordionItemProps`, which allows custom styling). This was due to time constraints - the current design doesn't seem to benefit much from this.
3. Accordion context provided via `EditableAccordion` 
4. Sidebar styling provided via `EditableSidebar` 

Closes [insert issue #]

## Notes
1. There is an existing `Card` component - but the abstraction we want (ability to toggle an expanded view) is an `Accordion`. separately, we should deprecate our `Card` component in favour of chakra's (newly introduced...)
2. form level stuff should be done automatically through `react-hook-form` rather than us having to manually juggle the error states + validation etc
3. should shift components out of `Editable` - will do so later
4. `SiteNotification` should be folded into `HeroSection` - will do later because it's not scoped for this PR

## Reviewer notes
1. Is it immediately clear what each component does? 
2. Do you feel that this mental model is sufficiently clear?

## Manual tests
- refer to #1393 - because _observed_ behaviour shouldn't change , the tests there are still valid here.

### CODE REVIEW

File: src/layouts/EditHomepage.jsx
Comment (seaerchin) [inline src/layouts/EditHomepage.jsx:1]: note to reviewers: the only change here is replacing the `Droppable` etc with `Editable.Draggable`. the large line diffs are due to indentation

**Review by alexanderleegs — COMMENTED** [review]
Just to check - the only `Editable` component being used currently is `Draggable` right, are the other components viewable anywhere?

File: src/layouts/components/Editable/Editable.tsx [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+export interface EditableDraggableProps {
+  onDragEnd: OnDragEndResponder
+  editableId: HomepageDroppableZones
+}
+export const EditableDraggable = ({
+  onDragEnd,
+  children,
+  editableId,
+}: PropsWithChildren<EditableDraggableProps>) => {
```
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/layouts/components/Editable/Editable.tsx:115]: yeap, will remove

File: src/layouts/components/Editable/Editable.tsx [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+  return (
+    // NOTE: According to the dnd docs,
+    // there CANNOT be more than 1
+    // `DragDropContext` in the component tree.
```
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/layouts/components/Editable/Editable.tsx:119]: if you have a `DragDropContext` as part of a component tree, there **cannot** be another `DragDropContext` in that same tree. 

concretely, this means that
```tsx
<DragDropContext>
    // some other components in between
    <DragDropContext>
    </DragDropContext>
</DragDropContext>
``` 
is not allowed. 

and wrt that - if you have an `Editable.Draggable` set up, it **must** wrap every _droppable_ item

File: src/layouts/components/Editable/Editable.tsx [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+  return (
+    <VStack spacing="0.5rem" alignItems="flex-start" px="1.5rem" pt="1.5rem">
+      <CustomiseSectionHeader />
+      <Flex
+        alignItems="center"
+        flexDir="column"
+        p="3.75rem 1.5rem"
+        justifyContent="center"
+      >
+        <HomepageStartEditingImage />
```
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/layouts/components/Editable/Editable.tsx:61]: could i clarify what you mean by this? in later PRs, the `EmptySidebar` is shifted out as a separate component and this component wraps the mapped components (ie, the infopic/bar/resources) in `EditHomepage`

**Comment by seaerchin** [comment]
> Just to check - the only `Editable` component being used currently is `Draggable` right, are the other components viewable anywhere?

done in next PR for ease of review

File: src/layouts/components/Editable/Editable.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/layouts/components/Editable/Editable.tsx]: removed all exports except `Editable` (namespace) + `CustomiseSectionHeader` (<- weird situation, doesn't feel like an `Editable` so i left it in)

File: src/assets/icons/BxDraggable.tsx [hunk 4]
```diff
@@ -0,0 +1,22 @@
+// eslint-disable-next-line import/prefer-default-export
```
Comment (kishore03109) [inline src/assets/icons/BxDraggable.tsx:1]: Hmm I think I made a pr to make this not throw an error already, am I mistaken here?

File: src/layouts/components/Editable/Editable.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (kishore03109) [reply src/layouts/components/Editable/Editable.tsx:119]: Hmm does the library throw an error if that is the case? 
if not do you think there is value to creating a custom Context here such that we warn the dev at runtime during local dev? That way we can reduce mental overhead when reasoning around this context

File: src/layouts/components/Editable/Editable.tsx [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+interface DraggableAccordionItemProps {
+  tag: JSX.Element
+  title: string
+  // TODO: Should get these props automatically
+  // rather than having us pass in manually
+  index: number
+  draggableId: string
+}
+// NOTE: Separating editable/draggable
```
Comment (kishore03109) [inline src/layouts/components/Editable/Editable.tsx:184]: may I clarify that all editable is a dragable?

File: src/layouts/components/Editable/Editable.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/layouts/components/Editable/Editable.tsx:119]: i haven't tried but i assume so. re: context - might do iff lib doesn't crash

File: src/assets/icons/BxDraggable.tsx (hunk 4 above)
Comment (seaerchin) [reply src/assets/icons/BxDraggable.tsx:1]: i just copy pasted so you might be right here? i can remove

File: src/layouts/components/Editable/Editable.tsx [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+      </AccordionButton>
+      <AccordionPanel pb={4}>{children}</AccordionPanel>
+    </BaseAccordionItem>
+  )
+}
+
+interface DraggableAccordionItemProps {
+  tag: JSX.Element
+  title: string
+  // TODO: Should get these props automatically
```
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/layouts/components/Editable/Editable.tsx:179]: yeap, and also the fact that we have to cast alot in `useDrag` suggests that the base abstraction should be on the array being drag/dropped and an interface should exist over it to have interop with stuff like this

File: src/layouts/components/Editable/Editable.tsx (hunk 5 above)
Comment (seaerchin) [reply src/layouts/components/Editable/Editable.tsx:184]: no - this is for accordion items only. editable accordion item (only 1 eg -  the homepage) don't have dragging capabilities + diff styling. draggable accordion items can be dragged.

File: src/styles/isomer-cms/pages/Editor.module.scss [hunk 7]
```diff
@@ -93,7 +93,6 @@
   box-shadow: 0 1px 4px 0 rgba(0, 0, 0, 0.2);
   overflow-y: scroll;
   box-sizing: border-box;
-  padding: 30px;
```
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/styles/isomer-cms/pages/Editor.module.scss:96]: not template - sidebar stuff. just realised this also impacts other pages so i'll add it back in at the cleanup PR (as the sidebar has to be totally removed)

File: src/layouts/EditHomepage.jsx [hunk 8]
```diff
@@ -975,197 +977,173 @@ const EditHomepage = ({ match }) => {
...
-                    </Droppable>
-                  </DragDropContext>
+                  <Editable.Draggable
+                    editableId="leftPane"
+                    onDragEnd={onDragEnd}
+                  >
+                    {frontMatter.sections.map((section, sectionIndex) => (
+                      <>
+                        {/* Hero section */}
+                        {section.hero ? (
```
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/layouts/EditHomepage.jsx:987]: done in a PR later on (can't remember which)

[the remark this answers is not in the document]
Comment (seaerchin) [reply src/layouts/EditHomepage.jsx:987]: also this was existing behaviour HAHA

File: src/layouts/components/Editable/Editable.tsx [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+    <VStack spacing="0.5rem" alignItems="flex-start" px="1.5rem" pt="1.5rem">
+      <CustomiseSectionHeader />
+      <Flex
+        alignItems="center"
+        flexDir="column"
+        p="3.75rem 1.5rem"
+        justifyContent="center"
+      >
+        <HomepageStartEditingImage />
+        <Text> Sections you add will appear here</Text>
```
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/layouts/components/Editable/Editable.tsx:62]: oops, no - wil remove

File: src/layouts/components/Editable/Editable.tsx [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+  Droppable,
+  DragDropContext,
+  Draggable,
+} from "@hello-pangea/dnd"
+import { IconButton } from "@opengovsg/design-system-react"
+import { PropsWithChildren } from "react"
+
+import { BxDraggable, HomepageStartEditingImage } from "assets"
+
+type SidebarHeaderProps = Pick<EditableSidebarProps, "title">
```
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/layouts/components/Editable/Editable.tsx]: oops, mutual reference @__@ fixed!

File: src/layouts/components/Editable/Editable.tsx [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+    >
+      <Text textStyle="subhead-3" color="base.content.brand">
+        {title}
+      </Text>
+    </Flex>
+  )
+}
+
+// TODO: Need to place this on top of draggable section
+export const CustomiseSectionHeader = () => (
```
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/layouts/components/Editable/Editable.tsx]: updated!

File: src/layouts/components/Editable/Editable.tsx [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+}: PropsWithChildren<EditableSidebarProps>) => {
+  return (
+    <Flex flexDir="column" bg="base.canvas.alt">
+      <SidebarHeader title={title} />
+      {children ? <>{children}</> : <EmptySideBarBody />}
+    </Flex>
+  )
+}
+
+type HomepageDroppableZones = "dropdownelem" | "leftPane" | "highlight"
```
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/layouts/components/Editable/Editable.tsx]: updated

File: src/layouts/components/Editable/Editable.tsx [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+  draggableId,
+}: PropsWithChildren<DraggableAccordionItemProps>) => {
+  return (
+    <Draggable draggableId={draggableId} index={index}>
+      {(draggableProvided) => (
+        <BaseAccordionItem
+          borderRadius="0.25rem"
+          {...draggableProvided.draggableProps}
+          ref={draggableProvided.innerRef}
+          // TODO: Ask design for the design token for this rgba value...
```
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/layouts/components/Editable/Editable.tsx:200]: asked! waiting on reply

**Review by kishore03109 — APPROVED** [review]

**Review by kishore03109 — COMMENTED** [review]
~~the usage of index arrays as keys is worrisome -> could we use some other unique id instead?~~ eh sial la wrong pr, somehow i cannot delete thi, can ignore

File: src/layouts/components/Editable/Editable.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/layouts/components/Editable/Editable.tsx:63]: This is regarding the `Homepage` specific component here!

[the remark this answers is not in the document]
Comment (seaerchin) [reply src/layouts/components/Editable/Editable.tsx:63]: not an issue as mentioned eawrlier, the `EmptySidebar` is shifted out alr so it's up to the caller to pass in the stuff

File: src/layouts/components/Editable/Editable.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/layouts/components/Editable/Editable.tsx]: fixed

**Pull request merged**